scintillometry/utils/pint_utils.py

```diff
@@ -0,0 +1,169 @@
+# Licensed under the GPLv3 - see LICENSE
+
+import warnings
+
+import numpy as np
+import astropy.units as u
+import astropy.time as time
+import pint.toa as toa
+import pint.models as model
+from pint import pulsar_mjd
+
+
+__all__ = ['PintUtils', 'make_toa_list']
+PINT_par_map = {'ephem': 'EPHEM', 'bipm_version': 'CLK',
+                'planets': 'PLANET_SHAPIRO'}
+
+
+class PintUtils(object):
+    """This is an utility class to compute the pulse phase at given time using
+       PINT (https://github.com/nanograv/PINT).
+
+       Parameter
+       ---------
+       par_file : str, optional
+           Tempo style parameter file
+       tim_file : str, optional
+           Tempo style TOA file. Default is None.
+       ephem : str, optinal
+           Solar system dynamic model file. Default is JPL 'de436'.
+       include_bipm : bool, optional
+           Flag to include the TT BIPM correction. Default is True.
+       bipm_version : str, optional
+           TT BIPM version. Default is 'BIPM2015'
+       include_gps : bool, optional
+           Flag to include the gps clock correction. Default is True.
+       planets : bool, optional
+           Flag to compute the planets' positions and velocities. Default is
+           False.
+       tdb_method : str, optional
+           The method to compute the TDB time scale. Default is using astropy
+           time objects' method.
+    """
+    def __init__(self, par_file=None, tim_file=None, ephem='de436',
+                 include_bipm=True, bipm_version='BIPM2015', include_gps=True,
+                 planets=False, tdb_method="default", **kwargs):
+        # Load model file
+        self.par_file = par_file
+        self.control_params = {'ephem': ephem, 'bipm_version': bipm_version,
+                               'include_bipm': include_bipm,
+                               'bipm_version': bipm_version,
+                               'include_gps': include_gps, 'planets': planets,
+                               'tdb_method': tdb_method}
+        self.control_params.update(kwargs)
+        # Load toa file
+        self.tim_file = tim_file
+
+    @property
+    def par_file(self):
+        return self._par_file
+
+    @par_file.setter
```

Comment:
I guess you have the setter to validate the parameters (at least, I'm not sure why we would allow after-the-fact setting), but I find it confusing that setting `par_file` ends up producing a `model` attribute. My sense would be to just do this in `__init__`.

Reply:
Maybe I was thinking from the point of taking advantage of PINT when I am writing this part. Sometimes, we want to use a new model for the same set of data. Reading a lot of TOAs is actually time-consuming.  That is why the after-the-fact setting is enabled.

Reply:
Yes, I can see that point. But I think we then need to think a bit about how to make the interface similar to `Predictor` (may well be easiest to adapt predictor, not at all sold on it!).

Reply:
I think I can change the architecture as follows:
```
pint_modle = PintModel(parfile, control_param=control_parameters)
pint_modle includes all the functions (phase, delay,...)
phase = pint_model.phase(times, or tim_files)
```
I think this makes `PintUtils` kind of like `Predictor`